# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

In [22]:
def create_pivot_table(data, group_by_column, agg_column='debt'):
    """
    Функция для создания сводной таблицы с количеством должников и соотношением должников.

    data: Данные
    group_by_column: Столбец, по которому нужно сгруппировать данные
    agg_column: Столбец, который будет агрегироваться (по умолчанию 'debt')
    return: Сводная таблица
    """
    # Сводная таблица с количеством должников
    pivot_table = data.groupby(group_by_column).agg(
        number_of_debtors=('debt', 'sum'),  # Количество должников
        total_people=('debt', 'count')       # Общее количество людей в группе
    )

    # Вычисляем долю должников
    pivot_table['proportion_of_debtors'] = pivot_table['number_of_debtors'] / pivot_table['total_people']

    # Форматируем вывод
    pivot_table['proportion_of_debtors'] = pivot_table['proportion_of_debtors'].apply(lambda x: f"{x:.2%}")  # Форматируем как процент

    print(f'Сводная таблица для {group_by_column}:')
    print(pivot_table)
    
    return pivot_table

In [23]:
def create_pivot_table(data, group_by_column, agg_column='debt'):
    """
    Функция для создания сводной таблицы с количеством должников и долей должников.

    data: Данные
    group_by_column: Столбец для группировки данных
    agg_column: Столбец для агрегации (по умолчанию 'debt')
    return: Сводная таблица
    """
    # Создаем сводную таблицу с количеством должников и долей должников
    pivot_table = data.groupby(group_by_column)[agg_column].agg(
        number_of_debtors='sum',             # Количество должников
        total_people='count',                # Общее количество в группе
        proportion_of_debtors='mean'         # Доля должников
    )
    
    # Форматируем долю должников как процент
    pivot_table['proportion_of_debtors'] = pivot_table['proportion_of_debtors'].apply(lambda x: f"{x:.2%}")
    
    # Отображаем таблицу
    display(pivot_table)
    
    return pivot_table

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [24]:
children_debt = create_pivot_table(data, 'children')  # сводная таблица по количеству детей

number_of_debtors  total_people proportion_of_debtors
children                                                       
0                      1063         14091                 7.54%
1                       444          4808                 9.23%
2                       194          2052                 9.45%
3                        27           330                 8.18%
4                         4            41                 9.76%
5                         0             9                 0.00%

**Вывод:** 
    между количеством детей и возвратом кредита в срок наблюдается зависимость. Процент просрочки кредитов увеличивается с увеличением количества детей (с 7.54% для заемщиков без детей до 9.45% для заемщиков с двумя детьми). Категории с 3, 4 и 5-ю детьми слишком малы, чтобы делать по нм выводы

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [25]:
family_status_debt = create_pivot_table(data, 'family_status')  # сводная таблица по семейному положению

number_of_debtors  total_people proportion_of_debtors
family_status                                                               
Не женат / не замужем                273          2796                 9.76%
в разводе                             84          1189                 7.06%
вдовец / вдова                        63           951                 6.62%
гражданский брак                     385          4134                 9.31%
женат / замужем                      927         12261                 7.56%

**Вывод:** 
Доля должников среди людей никогда не находившихся в официальном браке ("Не женат / не замужем", "гражданский брак") выше, чем в других группах 

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [26]:
total_income_debt = create_pivot_table(data, 'total_income_category')  #сводная таблица по категориям дохода

number_of_debtors  total_people proportion_of_debtors
total_income_category                                                       
A                                      2            25                 8.00%
B                                    354          5014                 7.06%
C                                   1353         15921                 8.50%
D                                     21           349                 6.02%
E                                      2            22                 9.09%

**Вывод:** 
Прямой зависимости между уровнем дохода и возвратом кредита в срок нет. Можно утверждать, что самыми дисциплинированными являются клиенты с доходом 30001 - 50000. Наибольший процент должников среди тех, чей уровень дохода менее 30000, однако в этой группе всего 22 человека получили кредит, поэту уверенно утврждать, что люди с самым низким доходом являются самыми недисциплинированными нельзя
Поскольку группы разделились не равномерно, делаю еще одну категоризацию по уровню дохода

In [27]:
def categorize_income_new(income):#определяю новую функцию для категоризации данных по доходу
    try:
        if 0 <= income <= 100000:
            return '<= 100000'
        elif 100001 <= income <= 150000:
            return '<= 150000'
        elif 150001 <= income <= 200000:
            return '<= 200000'
        elif income >= 200001:
            return '>= 200001'
    except:
        pass

In [28]:
data['total_income_category_new'] = data['total_income'].apply(categorize_income_new)

In [29]:
total_income_debt_new = create_pivot_table(data, 'total_income_category_new')  #новая сводная таблица по категориям дохода

number_of_debtors  total_people  \
total_income_category_new                                    
<= 100000                                354          4444   
<= 150000                                619          7110   
<= 200000                                403          4738   
>= 200001                                356          5039   

                          proportion_of_debtors  
total_income_category_new                        
<= 100000                                 7.97%  
<= 150000                                 8.71%  
<= 200000                                 8.51%  
>= 200001                                 7.06%

**Вывод:**
Процент просроченных кредитов имеет тенденцию к увеличению до уровня дохода 150000 и затем начинает снижаться. Процент должников наибольший для категории дохода от 100000 до 150000 (8.71%). Наименьший процент должников наблюдается в группе с доходом более 200000 (7.06%)

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [30]:
purpose_debt = create_pivot_table(data, 'purpose_category')  #сводная таблица по категориям назначения кредита

number_of_debtors  total_people  \
purpose_category                                            
операции с автомобилем                  400          4279   
операции с недвижимостью                780         10751   
получение образования                   369          3988   
проведение свадьбы                      183          2313   

                         proportion_of_debtors  
purpose_category                                
операции с автомобилем                   9.35%  
операции с недвижимостью                 7.26%  
получение образования                    9.25%  
проведение свадьбы                       7.91%

**Вывод:** 
Наименьшее количество должников среди людей, которые брали кредит на цели связанные с оперциями с недвижимостью. Наибольший процент должников среди людей с кредитами, связанными с автомобилем и получением образования

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
Пропуски в исходных данных могут возникнуть по следующим причинам:
1. Неполное заполнение анкет или форм – респонденты могли пропустить некоторые вопросы.
2. Технические сбои при проведении опросов или сборе данных (например, сбои в программном обеспечении или оборудовании).
3. Данные могли быть неправильно введены вручную, что привело к пропускам.
4. Ошибки форматирования, когда система не распознала данные и пропустила их.
5. Данные могут быть загружены из разных источников, и если некоторые источники имеют неполные записи, это может привести к пропускам.
6. Проблемы с совместимостью форматов данных могут вызвать потери или пропуски.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:*
Медиана, в отличие от среднего арифметического, менее чувствительна к выбросам. Если в данных содержатся экстремальные значения, среднее может быть сильно искажено, что приведет к неправильному заполнению пропусков. Медиана, будучи вычисленной как значение, разделяющее данные пополам, обеспечивает более стабильный и реалистичный подход.

### Шаг 4: общий вывод.

В ходе работы над проектом в данные были внесены следующие **изменения**:
1. Пропуски в столбцах days_employed и total_income заполнены медианными значениями;
2. В столбце children были удалены аномальные значения;
3. Созданы дополнительные столбцы total_income_category, total_income_category_new и purpose_category в которых все заемщики разделены на категории по уровню дохода и цели кредита

После анализа данных таблицы были сделаны следующие **выводы**: 
1. Между количеством детей и возвратом кредита в срок наблюдается зависимость. Процент просрочки кредитов увеличивается с увеличением количества детей (с 7.54% для заемщиков без детей до 9.45% для заемщиков с двумя детьми).
2. Доля должников среди людей никогда не находившихся в официальном браке ("Не женат / не замужем", "гражданский брак") выше, чем в других группах.
3. Процент просроченных кредитов имеет тенденцию к увеличению до уровня дохода 150000 и затем начинает снижаться. Процент должников наибольший для категории дохода от 100000 до 150000 (8.71%). Наименьший процент должников наблюдается в группе с доходом более 200000 (7.06%).
4. Наименьшее количество должников среди людей, которые брали кредит на цели связанные с оперциями с недвижимостью (7.26%). Наибольший процент должников среди людей с кредитами, связанными с автомобилем и получением образования (по 9%).

**ЗАКЛЮЧЕНИЕ**

Наибольшее количество должников наблюдается среди клиентов, которые никогда не были в официальном браке, имеют детей и доход ниже 30 000, и цель займа — приобретение автомобиля или получение образования. Так называемый "идеальный" клиент выглядит следующим образом: вдовец или вдова без детей с уровнем дохода от 200 001 до 1 000 000, целью кредита которого являются операции с недвижимостью.